design/design-doc-composition-functions.md

```diff
@@ -209,570 +738,353 @@ metadata:
-function must include the accumulated desired state in its output. Desired state
-is treated as an overlay on observed state, so a function pipeline need not
-specify the desired state of the XR (for example) unless a function wishes to
-mutate it.
-
-A full `FunctionIO` specification will accompany the implementation. Some
-example scenarios are illustrated below.
-
-A function that wanted to create (compose) a `CloudSQLInstance` would do so by
-returning the following `FunctionIO`:
-
-```yaml
-apiVersion: apiextensions.crossplane.io/v1alpha1
-kind: FunctionIO
-observed: {}  # Omitted for brevity.
-desired:
-  resources:
-  - name: cloudsqlinstance
-    resource:
-      apiVersion: database.gcp.crossplane.io/v1beta1
-      kind: CloudSQLInstance
-      spec:
-        forProvider:
-          databaseVersion: POSTGRES_9_6
-          region: us-central1
-          settings:
-            tier: db-custom-1-3840
-            dataDiskType: PD_SSD
-            dataDiskSizeGb: 20
-        writeConnectionSecretToRef:
-          namespace: crossplane-system
-          name: cloudsqlpostgresql-conn
-    connectionDetails:
-    - name: hostname
-      fromConnectionSecretKey: hostname
-    readinessChecks:
-    - type: None
+Starlark is a (very) limited dialect of Python designed for configuration. It
+embeds into many languages, including Go. This makes it a great candidate to
+build a generic Function that allows Composition authors to provide _inline_
+logic in a general-purpose-ish programming language. This is similar to the
+["GitHub Script"][github-script] GitHub Action that lets you plug some arbitrary
+logic into a GitHub Action pipeline when no existing Action does quite what you
+need.
+
+#### Iterating on Compositions
+
+Getting fast feedback on a Crossplane Composition has historically been a pain
+because the Composition logic has been encapsulated in a Kubernetes controller.
+In order to know whether your Composition works as intended you need to:
+
+1. Deploy (the right version of!) Crossplane to a Kubernetes cluster.
+2. Install Providers and Configurations for any resources you want to compose.
+3. Apply your XRD and Composition (potentially via a Configuration).
+4. Create an XR (or claim).
+5. See whether it works.
+
+Moving Composition logic out of Crossplane and into versioned Functions makes it
+a lot easier to test and iterate on, client-side. For example a `function test`
+command could test a single Function in isolation:
+
+```shell
+# Test a single Function by passing it a JSON RunFunctionRequest. This pulls and
+# starts the Function, makes the request, stops it, then returns the result.
+$ kubectl crossplane function test xpkg.upbound.io/negz/go-templates:v0.1.0 run-function-request.json
+{
+  # JSON encoded RunFunctionResponse omitted for brevity.
+}
 ```
 
-A function that wanted to set only an XR connection detail could return:
+You could similarly imagine a more holistic, `helm template`-like experience
```

Comment:
absolutely love this thinking here to drastically speed up the development process! 💪